### PR TITLE
better documentation for the best-fit allocation policy

### PR DESCRIPTION
**Pull request by gasche**
Currently the manual for 4.10 contains a documentation of the major heap allocation policies that is out-of-date, as it does not mention the shiny new best-fit policy of #8809 (the gc.mli documentation does).  (See the identical [manual section for 4.09](http://caml.inria.fr/pub/docs/manual-ocaml-4.09/runtime.html#ocamlrun-options))

This PR fixes this error by removing the hard-coded list of allocation policies in the manual, and redirecting to the documentation of the Gc module instead (it is only mentioned in text form, but a link is present just above in the OCAMLRUNPARAM documentation header).

I also improved the `allocation_policy` in gc.mli, by adding a note on the impact of less-fragmentation policies on space_overhead tuning. I don't understand the details of this fully, and @lpw25 and @damiendoligez gave slightly different explanations for the difference in space_overhead behavior (Leo says: wasted-space calculation ignores fragmentation; Damien says: defragmentation results in a small heap), so a second read would be welcome.

This should go in 4.10 when it's ready.

(cc @lpw25 @damiendoligez @Octachron)

### CODE REVIEW

File: stdlib/gc.mli [hunk 1]
```diff
@@ -145,17 +145,37 @@ type control =
...
+    (** The policy used for allocating in the major heap.
+        Possible values are 0, 1 and 2.
+        - 0 is the next-fit policy, which is usually fast
+          but can result in fragmentation.
+        - 1 is the first-fit policy, which avoids fragmentation
+          but is sensibly slower in some real-world use-cases
+          (it is only useful on programs exhibiting fragmentation issues
+           with next-fit)
+        - 2 is the best-fit policy, which is better than best-fit
+          and first-fit: it is both fast at allocating and excellent
```
Comment (gadmm) [inline stdlib/gc.mli]: I suggest a wording along the lines of replacing “... which is better than ...” with “... which empirical results suggests it is better than ...” otherwise it still sounds like this is a theoretical result.

Also, “... (since 4.10)”.

Comment (let-def) [reply stdlib/gc.mli]: Also typo: the second best-fit should be replaced by next-fit.

Comment (gasche) [reply stdlib/gc.mli]: Thanks to you both, I improved the text. I also added a small paragraph to reproduce some actual performance numbers, to give an idea of what kind of empirical differences we are talking about.

**Comment by Octachron** [comment]
We should also update `man/ocamlrun` before forgetting it.

File: manual/manual/cmds/runtime.etex [hunk 2]
```diff
@@ -126,14 +126,13 @@ The following environment variables are also consulted:
         This option takes no argument.
   \item[h] The initial size of the major heap (in words).
   \item[a] ("allocation_policy") The policy used for allocating in the
-  OCaml heap.  Possible values are 0 for the next-fit policy, and 1
-  for the first-fit policy.  Next-fit is usually faster, but first-fit
-  is better for avoiding fragmentation and the associated heap
-  compactions.
+    major heap. See the Gc module documentation for details.
```
Comment (Octachron) [inline manual/manual/cmds/runtime.etex]: I think that we should at least enumerate the possible values for the option, before pointing to the Gc module documentation.

Comment (gasche) [reply manual/manual/cmds/runtime.etex]: The slightly annoying thing with this choice is that it needs to be updated if we add a new policy again. Given the very small amount of details for most of the other option, I think that it's easier to just refer to the Gc manual (which surely people remember to update when they change the Gc?).

**Comment by gasche** [comment]
Actually Damien, being a man of his time, did think of updating `man/ocamlrun.m` in the original PR, he only forgot about the manual. The `man` page is as follows:

```
.BR a \ (allocation_policy)
The policy used for allocating in the OCaml heap.  Possible values
are 0 for the next-fit policy, 1 for the first-fit
policy, and 2 for the best-fit policy. Best-fit is still experimental,
but probably the best of the three. The default is 0.
```

We could leave it as-is -- and maybe Florian wants to argue for putting a similar amount of detail in the manual? -- or remove the duplicate content and refer to the Gc documentation again. (Left to myself I would just keep it as-is.)

File: stdlib/gc.mli [hunk 3]
```diff
@@ -145,17 +145,46 @@ type control =
...
+          result in fragmentation.
+
+        - 1 is the first-fit policy, which avoids fragmentation but
+          is sensibly slower in some real-world use-cases (it is only
+          useful on programs exhibiting fragmentation issues with
+          next-fit)
+
+        - 2 is the best-fit policy, which is empirically better than
+          next-fit and first-fit: in our experiments it is faster and
+          uses less memory than both.
```
Comment (Octachron) [inline stdlib/gc.mli]: `emprically better` sounds redundant with `in our experiments`. Also, the example below doesn't support the assertion
>it is faster and uses less memory than both.

 since best-fit is not strictly faster than next-fit and don't use strictly less memory than `first-fit`. Maybe, we could rephrase it to something like

> it combines the speed of next-fit with the fragmentation of first-fit

Comment (gasche) [reply stdlib/gc.mli]: Damien's number do show that best-fit tends to be both faster and less-memory-hungry than either modes, although that doesn't really show here with just this one run. (Note in particular that given the tradeoff between reclamation frequency and throughput, it doesn't really make sense to say that one strategy is "faster than first-fit but uses just as much memory" . For most workflows you can trade speed for space and get something that is slightly slower than before (but still faster than first-fit) and does use less memory.) Let me just replace the numbers I gave with a different run where the result is closer to the impression I want to give :-)

File: stdlib/gc.mli [hunk 4]
```diff
@@ -145,17 +145,46 @@ type control =
...
+
+        On one example that was known to be bad for next-fit and first-fit,
+        next-fit takes 28s using 855Mio of memory,
+        first-fit takes 47s using 566Mio of memory,
+        best-fit takes 27s using 545Mio of memory.
+
+        Note: the difference in fragmentation behavior means that
+        different policies will have different proportion of "wasted
+        space" for a given program: less compaction means a smaller
+        heap, so (for the same total memory consumption) more wasted
```
Comment (Octachron) [inline stdlib/gc.mli]: tuned ⇒ is tuned

    less compaction means a smaller heap

This is confusing: I am not sure to follow how less compaction imply a smaller heap?

File: manual/manual/cmds/runtime.etex (hunk 2 above)
Comment (Octachron) [reply manual/manual/cmds/runtime.etex]: Updating this list for every new policies doesn't sound like a gruesome burden? And I think it is better if the reader has at least the information about which values are possible before following the link.

File: stdlib/gc.mli (hunk 3 above)
Comment (Octachron) [reply stdlib/gc.mli]: Updating the example to be coherent with the claims is fine with me.

File: stdlib/gc.mli (hunk 4 above)
Comment (gasche) [reply stdlib/gc.mli]: "less fragmentation", sorry, will fix

File: manual/manual/cmds/runtime.etex (hunk 2 above)
Comment (gasche) [reply manual/manual/cmds/runtime.etex]: Fine, I reused the manpage's description. (I tried to just enumerate the possible values, but it just looked like bad documentation.)

Comment (Octachron) [reply manual/manual/cmds/runtime.etex]: Thanks!

File: stdlib/gc.mli [hunk 5]
```diff
@@ -145,17 +145,45 @@ type control =
...
+        best-fit takes 27s using 545Mio of memory.
+
+        Note: the difference in fragmentation behavior means that
+        different policies will have different proportion of "wasted
+        space" for a given program. Less fragmentation means a smaller
+        heap so, for the same amount of wasted space, a higher
+        proportion of wasted space, making the GC work harder. When
+        changing to a low-fragmentation policy, you may need to
+        augment the [space_overhead] setting, for example using [100]
+        instead of the default [80] which tuned for next-fit.
```
Comment (Octachron) [inline stdlib/gc.mli]: There is still `[80] which tuned` that looks like it should be `which is tuned`. I also wonder if the paragraph structure should be reversed: start with remarking that the `[space_overhead]` setting may need to be tuned differently for each policies and then explain why this fine tuning is dependent on the policy.

Comment (gasche) [reply stdlib/gc.mli]: Nice suggestion, thanks! Applied.

**Comment by gasche** [comment]
@Octachron do I understand correctly that you are approvingly satisfied with the final state?

**Review by Octachron — APPROVED** [review]
I like the current text indeed.
